server/src/main/java/org/opensearch/search/sort/SortedWiderNumericSortField.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.search.sort;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.FieldComparator;
+import org.apache.lucene.search.LeafFieldComparator;
+import org.apache.lucene.search.SortedNumericSelector;
+import org.apache.lucene.search.SortedNumericSortField;
+import org.apache.lucene.search.comparators.NumericComparator;
+
+import java.io.IOException;
+
+/**
+ * Sorted numeric field for wider sort types,
+ * to help sorting two different numeric types.
+ *
+ * @opensearch.internal
+ */
+public class SortedWiderNumericSortField extends SortedNumericSortField {
```

Comment:
I am wondering why only for docs merge and not make this sort a default one since it (apparently) brings sizeable gains?

Reply:
Lucene sort optimization is only applicable to Numeric Types. [Reference](https://lucene.apache.org/core/9_0_0/core/org/apache/lucene/search/SortField.html#:~:text=the%20next%20release.-,setOptimizeSortWithPoints,-%40Deprecated%20public%C2%A0void).
We have enabled all NumericTypes with this PR for Sort optimization which we disabled previously last week with this PR [6321](https://github.com/opensearch-project/OpenSearch/pull/6321),

I think I misunderstood your question probably, since you already knew that. Are you saying we should enable Sort optimization while doing merging of shard results too ?

Reply:
> I think I misunderstood your question probably, since you already knew that. Are you saying we should enable Sort optimization while doing merging of shard results too ?

That's right, why we cannot promote these comparators to be the "default" ones for numeric types?

Reply:
Sort optimization will be enabled for all Numeric types now with this PR. This comparator SortedWiderNumericSortField is not contributing to improve optimization, but it is allowing us to sort/merge different types of numeric types (merging already srted docs). Which is needed to merge results from shard which have different SortField.Type.
Main contribution for this latency optimization is still Lecene Optimization which we disabled [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/fielddata/IndexNumericFieldData.java) for few types still, 

Lucene choses this Int/Long/Float/Double comparator accordingly with the sort type we have specified.

Reply:
For individual shards, we dont need this comparator at all, since in one shard, we will always have only one type of field mapping

Reply:
I see, thanks @gashutos !